File: flagPartition/partition.cc
#include <iostream> // cout
#include <stdlib.h> // rand srand
#include <time.h> // time

using namespace std;

int partition(int array[], int n, int* left, int* right)
{
    if(!array || !n)
    {
        return 0;
    }

    int small = array[0];
    int large = array[n-1];
    if(small > large)
    {
        int tmp = large;
        large = small;
        small = tmp;
    }

    int wr = n-1;
    int wl = 0;

    for(int reader = 0; reader <= wr; )
    {
        int v = array[reader];

        if(v < small)
        {
            array[reader] = array[wl];
            array[wl] = v;
            wl++;
        }
        else if(v > large)
        {
            array[reader] = array[wr];
            array[wr] = v;
            wr--;
        }
        else
        {
            reader++;
        }
    }
    *left = wl;
    *right = wr;

    return 1;
}

void print(int array[], int n)
{
    for(int i=0; i < n; i++)
    {
        cout << array[i] << " ";
    }
    cout << "\n";
}

int main()
{
    int n = 10;
    int b[n];
    srand(time(NULL));

    int start = clock();
    for(int j = 0; j < 1000 * 1000; j++)
    {
        for(int i = 0; i < n; i++)
        {
            b[i] = rand() % n;
        }
        // print(b, n);
        int l, r;
        partition(b, n, &l, &r);
        // print(b, n);
    }
    int end = clock();
    cout << ((float)end - start)/CLOCKS_PER_SEC;

    // cout << l << " " << r;
}

File: flagPartition/partition.py
import random
import sys
import time

def partition(array):
    if array is None or len(array) == 0:
        raise TypeError

    small = array[0]
    large = array[len(array)-1]

    if small > large:
        small, large = large, small

    wr = len(array) - 1
    wl = 0
    reader = 0

    while reader <= wr:
        v = array[reader]

        if v < small:
            array[wl], array[reader] = v, array[wl]
            wl += 1
        elif v > large:
            array[wr], array[reader] = v, array[wr]
            wr -= 1
        else:
            reader += 1

    return (wl, wr)

def test(array, fr, to):
    if len(array) == 1 and fr == to:
        return

    i = 0
    min = sys.maxint if fr > 0 else -sys.maxint
    while i < fr:
        if array[i] < min:
            min = array[i]
        i += 1

    i = to+1
    max = -sys.maxint if fr == len(array) -1 else sys.maxint
    while i < len(array):
        if array[i] > max:
            max = array[i]
        i += 1

    if min > max:
        raise StopIteration

    i = fr
    while i <= to:
        v = array[i]
        if v <= min or v >= max:
            raise StopIteration
        i += 1

    pass

arr = [1]
l, r = partition(arr)
test(arr, l, r) 

arr = [1,10]
l, r = partition(arr)
test(arr, l, r) 

arr = [1,10,4]
l, r = partition(arr)
test(arr, l, r) 

arr = [1,10,4,3,2,5,7,6,9,8]
l, r = partition(arr)
test(arr, l, r) 

arr = [10,1,4,3,2,5,7,6,9,8]
l, r = partition(arr)
test(arr, l, r) 

arr = [18, 23, 20, 91, 2]
l, r = partition(arr)
test(arr, l, r) 

tests = 100000
for j in range(tests):
    arr = []
    length = random.randint(3,10)
    for i in range(length):
        arr.append(random.randint(0,100))
    l, r = partition(arr)
    test(arr, l, r) 

print("Tested {0} random cases successfully.".format(tests))

start = time.time()

for j in range(1000 * 1000):
    arr = [0] * 10
    # arr = []
    for i in range(10):
        # arr[i] = random.randint(0,10) # 20 sec
        arr[i] = int(random.random()*10) # 9 sec
        # arr.append( random.randint(0,10) ) # 20 sec
        pass
    l, r = partition(arr)

end = time.time()

print(end - start)
